### ISSUE TITLE
What to do with 5 outputs from multiscale deep supervision in Unet3+ ?  fusion or take multi loss at each scale 

### ISSUE THREAD

**Issue by rammyram** [ZJUGiveLab/UNet-Version#7]
Should we concat output along channels dim and perform 
Conv2d(n_classes*5, n_classes, 1)  such that we finally have one output ? 

or 

They need to be trained independently, each having its loss like below ? 

# Clear gradients
optimizer.zero_grad()
o1,o2,o3,o4,o5 = model(data)

# Calc hybrid loss 
loss = 0 
for outval in [o1,o2,o3,o4,o5]:
  loss += get_hybrid_loss(outval,target)

# loss back
loss.backward()

but the training is really slow this way

**Comment by cholojuanito** [comment]
I had the same question. Seeing the code that the @ZJUGiveLab used to train the deep supervision network would be great!

**Comment by InvincibleXiao** [comment]
I had the same question.

**Comment by cholojuanito** [comment]
Actually, I figured it out you can just add all the losses together. I'm not sure why I thought I couldn't just add the losses up.

A loss is just a number that is part of the gradient. You can add those losses together and then call backward on the cumulative loss.

Here's some example code:

```
logits = network(inputs)

logit_loss_sum = 0
for logit in logits:
    logit_loss_sum += loss_func(logit, label)

logit_loss_sum.backward()
```
As far as performance goes I think this model is just slow, it might be the way that I built my version is slow but I also couldn't do more than a batch size of 1 because of GPU constraints

**Comment by InvincibleXiao** [comment]
Yes,the parameter of this network is very huge.This network has three loss functions, and I don't know which layer of output corresponds to which loss function.------------------&nbsp;原始邮件&nbsp;------------------
发件人:&nbsp;&quot;Tanner&amp;nbsp;Davis&quot;<notifications@github.com&gt;
发送时间:&nbsp;2021年2月9日(星期二) 晚上10:34
收件人:&nbsp;&quot;ZJUGiveLab/UNet-Version&quot;<UNet-Version@noreply.github.com&gt;;
抄送:&nbsp;&quot;InvincibleXiao&quot;<814785779@qq.com&gt;;&quot;Comment&quot;<comment@noreply.github.com&gt;;
主题:&nbsp;Re: [ZJUGiveLab/UNet-Version] What to do with 5 outputs from multiscale deep supervision in Unet3+ ?  fusion or take multi loss at each scale  (#7)

**Comment by cholojuanito** [comment]
> I don't know which layer of output corresponds to which loss function

The 3 loss function results are just added together.
You should just pass all the layers, one at a time, through the hybrid_loss function. All the logits are the same size as the original image that was passed into the network/model.

**Comment by InvincibleXiao** [comment]
I tried.But how to use MSSSIM(),the output is not a number

**Comment by rammyram** [comment]
Can't believe I missed this, just a simple concat would do, thanks @cholojuanito

**Comment by slience-ops** [comment]
the function 'get_hybrid_loss ' ，how to code?
